File: dynamic-programming/java/LongestCommonSubsequence.java
/**
* @author 	Luca Cortinovis	luca1cortinovis@gmail.com
* @version 	1.0
* @since	2/11/2022
*
* You can find me here:
* GitHub	Luca Cortinovis(NeptuneLuke)
*/


import java.util.Scanner;
import java.lang.Math;

public class LongestCommonSubsequence {

	
	public static void main(String args[]) {
		
		Scanner keyboard = new Scanner(System.in);
		
		System.out.println("Insert first sequence: ");
		String X = keyboard.next();
		int m = X.length();
		
		System.out.println("Insert second sequence: ");
		String Y = keyboard.next();
		int n = Y.length();
		
		int myMatrix[][] = LCS_length(X,Y);				// fill myMatrix with the lengths of the solutions of all subproblems
		
		String sol = LCS_backtrack(X,Y,m,n,myMatrix);	// build one possible solution from the length
		
		System.out.println("LCS length: " + sol.length());
		System.out.println("One possible LCS between X,Y: " + sol);
	}
	
	
	/* LCS algorithm for two generic sequences, this function calculates the length of the/a LCS between X,Y
	 * if i need to find the LCS for n sequences, just build the LCS from the first two, then find the LCS between that and the third
	 * and go on... 
	 * 
	 * Time complexity: O(m*n)
	 */
	public static int[][] LCS_length(String X, String Y) {
		
		int m = X.length();
		int n = Y.length();
		int C[][] = new int [m+1][n+1]; 	// matrix of the lengths of LCS 
		
		
		for(int i=0; i<=m; i++)		// base case -> solution: empty string
			C[i][0] = 0;
		
		for(int j=0; j<=n; j++)		// base case -> solution: empty string
			C[0][j] = 0;
		
		
		// "recursive" step
		for(int i=1; i<=m; i++) {
			for(int j=1; j<=n; j++) {
				
				if(X.charAt(i-1) == Y.charAt(j-1))		// charAt(index-1) because Strings start at 0
					C[i][j] = C[i-1][j-1] + 1;
				else
					C[i][j] = Math.max(C[i-1][j], C[i][j-1]);
			}
		}
		
		return C;
	}
	
	
	/* LCS_backtrack builds the actual LCS
	 * this type of algorithm builds ONLY ONE possible LCS (it can be more than one) */
	public static String LCS_backtrack(String X, String Y,int i, int j, int C[][]) {
		
		if(i == 0 || j == 0)
			return "";
		
		else {
			
			// if X,Y end with the same symbol go diagonal
			if(X.charAt(i-1) == Y.charAt(j-1))
				return LCS_backtrack(X,Y,i-1,j-1,C) + X.charAt(i-1);
			
			else {
				
				// if top > left then go top
				if(C[i-1][j] >= C[i][j-1])
					return LCS_backtrack(X,Y,i-1,j,C);
				
				else	// else go left 
					return LCS_backtrack(X,Y,i,j-1,C);
				
				
				/* this method finds another possible LCS
				 * 
				 * // if left > top then go left
				 * if(C[i][j-1] >= C[i-1][j])
						return LCS_backtrack(X,Y,i,j-1,C);
				   else		// else go top
						return LCS_backtrack(X,Y,i-1,j,C); 
				 */
			}
		}
		
	}
	
}

File: dynamic-programming/cpp/LongestCommonSubsequence.cpp
/**
    * @author 	Luca Cortinovis	luca1cortinovis@gmail.com
    * @since	28/10/2023
    *
    * You can find me here:
    * GitHub	Luca Cortinovis (NeptuneLuke)
*/

#include <iostream>
#include <string>
#include <vector>


using
std::string,
std::vector,
std::cout,
std::cin;


int lcs_length(string s1, string s2, vector< vector<int> > &lcs_table);
vector<char> lcs_build(string s1, string s2, vector <vector<int> > lcs_table);
string lcs_backtrack(string s1, string s2, int i, int j, vector< vector<int> > lcs_table);


int main(void) {

    string s1, s2;
    cout << "Insert the first string \n";
    getline(cin, s1);
    int m = s1.size();

    cout << "Insert the second string \n";
    getline(cin, s2);
    int n = s2.size();

    vector< vector<int> > lcs_table(m + 1, vector<int>(n + 1));


    // Calculate the length of the LCS
    int l = lcs_length(s1, s2, lcs_table);
    cout << "\n\n" << "Length of the LCS between " << s1 << " and " << s2 << " : " << l << "\n\n";


    // Build the actual LCS (method 1)
    vector<char> lcs = lcs_build(s1, s2, lcs_table);
    cout << "LCS length: " << lcs.size()-1 << "\n";
    cout << "The LCS between " << s1 << " and " << s2 << " : ";
    for (auto elem : lcs) {
        cout << elem << "";
    }
    cout << "\n\n";


    // Build the actual LCS (method 2)
    string s3 = lcs_backtrack(s1, s2, m, n, lcs_table);
    cout << "LCS length: " << s3.size() << "\n";
    cout << "The LCS between " << s1 << " and " << s2 << " : " << s3 << "\n\n";


    return 0;
}


int lcs_length(string s1, string s2, vector< vector<int> > &lcs_table) {

    int m = s1.size();
    int n = s2.size();

    for (int i = 0; i <= m; ++i) {

        for (int j = 0; j <= n; ++j) {

            // base case S1[i]==0 or S2[j]==0
            if (i == 0 || j == 0) {
                lcs_table[i][j] = 0;
            }

            // recursive step S1[i] == S2[j]
            else if (s1[i-1] == s2[j-1]) {
                lcs_table[i][j] = 1 + lcs_table[i-1][j-1];
            }

            // recursive step S1[i] != S2[j]
            else {
                lcs_table[i][j] = std::max(lcs_table[i-1][j], lcs_table[i][j-1]);
            }
        }
    }

    return lcs_table[m][n];
}


vector<char> lcs_build(string s1, string s2, vector <vector<int> > lcs_table) {

    int m = s1.size();
    int n = s2.size();

    int index = lcs_table[m][n];
    vector<char> lcs(index + 1);
    lcs[index] = '\0';

    int i = m, j = n;
    while (i > 0 && j > 0) {

        if (s1[i-1] == s2[j-1]) {
            lcs[index-1] = s1[i-1];
            i--;
            j--;
            index--;
        }
        else if (lcs_table[i-1][j] >= lcs_table[i][j-1])
            i--;
        else
            j--;
    }

    return lcs;
}


string lcs_backtrack(string s1, string s2, int i, int j, vector< vector<int> > length) {

    if (i == 0 || j == 0) {
        return "";
    }
    else {

        // if S1,S2 end with the same symbol go diagonal
        if (s1[i-1] == s2[j-1]) {
            return lcs_backtrack(s1, s2, i - 1, j - 1, length) += (s1[i-1]);
        }
        else {
            // if top > left then go top
            if (length[i-1][j] >= length[i][j-1])
                return lcs_backtrack(s1, s2, i - 1, j, length);

            else	// else go left 
                return lcs_backtrack(s1, s2, i, j - 1, length);
        }
    }
}
